src/Build/BackEnd/Components/Logging/LoggingService.cs

```diff
@@ -521,43 +521,64 @@ public bool HasBuildSubmissionLoggedErrors(int submissionId)
 
         /// <summary>
         /// Returns a hashset of warnings to be logged as errors for the specified build context.
-        /// Note that WarningsAsMessages takes priority over WarningsAsErrors and are excluded from the set.
-        ///
-        /// If all warnings to be treated as errors should also be treated as messages, return null.
-        /// This is to avoid all warnings being treated as errors. <see cref="RequestBuilder.ConfigureWarningsAsErrorsAndMessages()"/>
         /// </summary>
         /// <param name="context">The build context through which warnings will be logged as errors.</param>
         /// <returns>
-        /// An empty set if all warnings should be treated as errors.
-        /// A set containing warning codes to be logged as errors.
-        /// Null if no warnings should be treated as errors.
         /// </returns>
-        public ICollection<string> GetWarningsToBeLoggedAsErrorsByProject(BuildEventContext context)
+        public ICollection<string> GetWarningsAsErrors(BuildEventContext context)
         {
-            if (_warningsAsErrorsByProject == null)
+            int key = GetWarningsAsErrorOrMessageKey(context);
+
+            // If there is definitely nothing to convert into an error, return early.
+            if (WarningsAsErrors == null && (_warningsAsErrorsByProject == null || !_warningsAsErrorsByProject.ContainsKey(key)))
             {
                 return null;
             }
 
+            HashSet<string> allWarningsAsErrors = new HashSet<string>();
+
+            if (WarningsAsErrors != null)
+            {
+                allWarningsAsErrors.UnionWith(WarningsAsErrors);
+            }
+
+            if (_warningsAsErrorsByProject != null)
+            {
+                if (_warningsAsErrorsByProject.TryGetValue(key, out ISet<string> warningsAsErrors))
+                {
+                    allWarningsAsErrors.UnionWith(warningsAsErrors);
+                }
+            }
+
+            return allWarningsAsErrors;
```

Comment:
This could be simplified a good bit. Also, if they're both null, we should return null and not an empty HashSet, right? Otherwise, we'd be indicating that all warnings should be counted as errors. I'd recommend something similar below. Also, since they're essentially the same method, it seems a lot easier to make them one method and call it with slightly different parameters. (You don't have to change the public API for that, since you can make the heart of the implementation private.)
```suggestion
            ISet<string> warningsAsErrors = null;
            if (_warningsAsErrorsByProject is not null && _warningsAsErrorsByProject.TryGetValue(key, out warningsAsErrors)
            {
                    if (WarningsAsErrors is not null)
                    {
                            warningsAsErrors.UnionWith(WarningsAsErrors);
                    }
                    return warningsAsErrors;
            }
            else
            {
                    return WarningsAsErrors;
            }
```

Reply:
If they're both null they'll return through the first if statement. I do like your suggested change though, less to sift through.